### ISSUE TITLE
ttnn.rms_norm on a rank-0 tensor computes x / |x|: epsilon is dropped, x = 0 gives NaN and every |x| < sqrt(eps) gives ±1

### ISSUE THREAD

**Issue by ping-dev-ui** [tenstorrent/tt-metal#55347]
**Component / Area:** ttnn normalization, `rms_norm` host dispatch (rank-0 branch)

**Issue Type:** Bad Outputs

## Summary

`rms_norm` has an explicit rank-0 path
([rmsnorm.cpp#L43-L58](https://github.com/tenstorrent/tt-metal/blob/9fa00fce13cda1538f0130695576408ff5b05665/ttnn/cpp/ttnn/operations/normalization/rmsnorm/rmsnorm.cpp#L43-L58)):

```cpp
// For 0D tensors
if (rank == 0) [[unlikely]] {
    auto result = ttnn::divide(input_tensor, ttnn::abs(input_tensor, output_memory_config), ...);   // x / |x|
    if (weight) ... multiply;  if (bias) ... add;
    return result;
}
```

`epsilon` is never read on this branch; it is only forwarded to `prim::layer_norm` for rank >= 1. The registered golden
([normalization.py#L113-L125](https://github.com/tenstorrent/tt-metal/blob/9fa00fce13cda1538f0130695576408ff5b05665/ttnn/ttnn/operations/normalization.py#L113-L125))
is `x * rsqrt(mean(x^2) + eps)`, which for a scalar is `x / sqrt(x^2 + eps)`, and `torch.nn.functional.rms_norm` agrees. So the
op returns `0 / 0 = NaN` at `x = 0` and exactly `±1` for every `|x|` below about `sqrt(eps)`, where the reference is a value
between 0 and 1.

This is a source-level analysis with a host replay; I do not have a card. The composition is two elementwise ops, so the
device result is what the replay shows up to rounding.

## Observed vs expected (eps = 1e-5 unless noted)

| x | `rms_norm` now | golden / `F.rms_norm` |
|---|---|---|
| 0 | **NaN** | 0 |
| 1e-7 | **1** | 3.16e-5 |
| 1e-3 | **1** | 0.3015 |
| -2e-3 | **-1** | -0.5345 |
| 1e-2 | 1 | 0.9535 |
| 1e-6 (eps = 1e-12) | **1** | 0.7071 |
| 1, -5 | 1, -1 | 1, -1 |

## Why the tests do not catch it

The rms_norm tests exercise rank >= 1 tensors; the rank-0 branch has no test with a small or zero input.

Related but different: #49600 (fp32 accuracy of the rank >= 1 kernel).

## Reproducer (host only, `pip install torch`)

```python
import torch

def ttnn_rms_norm_0d(x, eps):                 # rmsnorm.cpp L45-49: epsilon unused
    return x / torch.abs(x)

def golden(x, eps):                           # normalization.py L113-122
    return x * torch.rsqrt(x.to(torch.float32).pow(2).mean(-1, keepdim=True) + eps)

for xv in (0.0, 1e-7, 1e-3, -2e-3):
    x = torch.tensor(xv)
    print(xv, ttnn_rms_norm_0d(x, 1e-5).item(), golden(x, 1e-5).item(),
          torch.nn.functional.rms_norm(x.reshape(1), [1], eps=1e-5).item())
# 0.0 nan 0.0 0.0
# 1e-07 1.0 3.16e-05 3.16e-05
# 0.001 1.0 0.3015 0.3015
# -0.002 -1.0 -0.5345 -0.5345
```

## Proposed fix

```cpp
// x * rsqrt(x^2 + eps): keeps epsilon and is finite at 0
auto result = ttnn::multiply(
    input_tensor,
    ttnn::rsqrt(ttnn::add(ttnn::square(input_tensor, output_memory_config), epsilon, std::nullopt, output_memory_config),
                output_memory_config),
    std::nullopt,
    output_memory_config);
```

or route rank-0 through the rank-1 kernel after a reshape to `{1}`. Add a rank-0 test with `x = 0` and `|x| < sqrt(eps)`.

## Acceptance criteria (proposed)

- Rank-0 `rms_norm` matches `torch.nn.functional.rms_norm` for `x` in `{0, ±1e-7, ±1e-3, ±1, ±5}` at eps `1e-5` and `1e-12`, fp32 and bf16, with and without weight/bias.
- Test added for the rank-0 branch.

**Comment by ping-dev-ui** [comment]
Reporter here. Fix with tests is ready on my fork (https://github.com/ping-dev-ui/tt-metal/tree/fix/rms-norm-rank0-eps): the rank-0 path computes x * rsqrt(x^2 + eps), the golden formula specialised to a scalar, so epsilon is kept, x = 0 gives 0 and small |x| no longer saturates to ±1; a rank-0 test covers x in {0, ±1e-7, 1e-3, -2e-3, 1e-2, 1, -5} at eps 1e-5 and 1e-12 in bf16 and fp32, with and without weight and bias. Holding the PR until assignment per the bounty terms. I have read and agree to the Tenstorrent Bounty Program Terms & Conditions (https://docs.tenstorrent.com/bounty_terms.html). @jberkowitzTT could you assign this to me when it is reviewed?

**Comment by heidman06** [comment]
Implemented in PR #55348 with exact scalar RMSNorm formulation incorporating epsilon and unit tests covering zero and sub-sqrt(eps) values against torch.nn.functional.rms_norm reference.

**Comment by jberkowitzTT** [comment]
Hi @heidman06  - please submit all bounty proposals to [bounties@tenstorrent.com](mailto:bounties@tenstorrent.com). If approved, we'll create the issue and assign it to you as a bounty

Thank you for your enthusiasm & understanding

**Comment by heidman06** [comment]
Thanks for the guidance @jberkowitzTT, proposal submitted to bounties@tenstorrent.com with the PR details!
